pubsub.go

```diff
@@ -73,7 +82,10 @@ type PubSub struct {
 	peerDead chan peer.ID
 
 	// The set of topics we are subscribed to
-	myTopics map[string]map[*Subscription]struct{}
+	mySubs map[string]map[*Subscription]struct{}
+
+	// The set of topics we are interested in
+	myTopics map[string]*Topic
```

Comment:
My understanding is that PubSub is supposed to be allowed to `Publish` without `Subscribing` (e.g. see the fanout logic in gossipsub which makes no sense otherwise). So we do in fact need to keep these things separate unless we change the PubSub semantics.

From a programming perspective I think it's totally reasonable to have a `subs map[*Subscription]struct{}` field in `Topic`, I just figured it would be easier to read the PR if I left it as it was. Happy to change it if you're game though (just give a 👍)

Reply:
Hm. Yeah, that does make sense. I'm a bit concerned that, if you aren't _subscribed_ to a topic, peers might cut the connection because they don't know you're publishing.

I _would_ like to add an option to `Join` for auto-subscribing (but throwing away the results) but let's punt on that.

Reply:
Makes sense, we'll punt for now. What you're recommending seems like a part of #28.